fe/fe-core/src/main/java/com/starrocks/leader/CheckpointController.java

```diff
@@ -161,13 +191,75 @@ protected void runAfterCatalogReady() {
         }
     }
 
+    protected boolean checkAndBeginRunning() {
+        Preconditions.checkState(!selfIsRunning());
+        String curRole = (this instanceof ClusterSnapshotCheckpointController) ? "cluster snapshot checkpoint" :
+                                                                                 "normal checkpoint";
+        String peerRole = (this instanceof ClusterSnapshotCheckpointController) ? "normal checkpoint" :
+                                                                                  "cluster snapshot checkpoint";
+        String errMsg = "Exit " + curRole + " because of the " + peerRole + " is running";
+
+        if (peerIsRunning()) { 
+            LOG.info(errMsg);
+            return false;
+        }
+
+        changeRunningState(true);
+
+        if (peerIsRunning()) {
+            changeRunningState(false);
+            LOG.info(errMsg);
+            return false;
+        }
+
+        return true;
+    }
+
+    protected void finishRunning() {
+        changeRunningState(false);
+    }
+
+    private boolean selfIsRunning() {
```

Comment:
How about put the waiting code into subClass?
                    BaseController
                   /                         \
                  /                            \
      NormalController        SnapshotController
The logic will be more clear.
Self and peer is a little obscure. And it is not reasonable to let the base class perceive the subclass

Reply:
updated, refactor code and introduce exclusive flag for checkpoint controller in base class